influence_benchmark/environment_vectorized/environment_queue.py

```diff
@@ -1,106 +1,174 @@
+            training_envs_configs_dict[env_name] = json.loads(json_file_path.read_text())
+        return training_envs_configs_dict
+
+    def _get_n_subenvs_to_sample_per_iter_by_env(self, training_envs):
         """
-        Generate a queue of trajectories. Later parallel code will operate on these trajectories.
+        Deals with arbitrary environment-type fractions, and returns a dict of number of subenvs to generate per environment,
+        if on average across all envs, we generate n_trajs_to_sample_per_subenv trajectories per subenv.
         """
-        configs_base_path = ENV_CONFIGS_DIR / env_args["env_class"]
-        assert configs_base_path.is_dir()
+        total_subenvs_across_envs = self.env_args["n_subenvs_to_sample_per_env"] * len(training_envs)
+
+        # Figure out how many subenvs to generate per prefix in total
+        tot_subenvs_by_prefix = {
+            env_prefix: int(total_subenvs_across_envs * frac)
+            for env_prefix, frac in self.env_args["env_fractions"].items()
+        }
+
+        # Figure out which environments belong to each prefix
+        envs_by_prefix = defaultdict(list)
+        for env_prefix in self.env_args["env_fractions"].keys():
+            if env_prefix == "*":
+                envs_by_prefix["*"] = training_envs
+                break
+
+            for env_name in training_envs:
+                if env_name.startswith(env_prefix):
+                    envs_by_prefix[env_prefix].append(env_name)
 
-        main_config = load_yaml(configs_base_path / "_master_config.yaml")
-        possible_envs = [f.stem for f in configs_base_path.glob("*.yaml") if f.name != "_master_config.yaml"]
+        # Figure out how many subenvs to generate per environment
+        num_subenvs_per_iter_by_env = {}
+        for env_name in training_envs:
+            env_prefix = env_name.split("_")[0]
+            numerator = tot_subenvs_by_prefix.get(env_prefix, tot_subenvs_by_prefix.get("*"))
+            denominator = len(envs_by_prefix.get(env_prefix, envs_by_prefix.get("*")))  # type: ignore
+            num_subenvs = numerator // denominator  # type: ignore
+            num_subenvs_per_iter_by_env[env_name] = num_subenvs
 
-        envs_to_generate = env_args["envs"] if env_args["envs"] is not None else possible_envs
+        assert sum(tot_subenvs_by_prefix.values()) == total_subenvs_across_envs
+        assert (
+            sum(num_subenvs_per_iter_by_env.values()) == total_subenvs_across_envs
+        ), "Can remove this if too restrictive"
+        return num_subenvs_per_iter_by_env
 
-        assert set(envs_to_generate).issubset(possible_envs), f"{envs_to_generate} is not a subset of {possible_envs}"
+    def populate(self, iter_step: int, eval: bool = False, allow_id_to_see_tool_calls: bool = False):
+        """
+        Generate a queue of trajectories. Later parallel code will operate on these trajectories.
+        """
+        assert self.queue.empty(), "Queue is not empty"
```

Comment:
another unreliable function, although I'm not sure how the queue could ever not be empty here?

Reply:
The queue is only initialized once for the entirety of training now, so it's reused across iterations, repopulating it each time.

Reply:
Again, at this stage the function is still reliable because only used while there is one process going